kliff/transforms/configuration_transforms/descriptors.py

```diff
@@ -0,0 +1,477 @@
+
+
+@requires(lds, "libdescriptor is needed for Descriptors")
+def show_available_descriptors():
+    """
+    Show all the available descriptors in libdescriptor.
+    """
+    print("-" * 80)
+    print(
+        "Descriptors below are currently available, select them by `descriptor: str` attribute:"
+    )
+    print("-" * 80)
+    _instance = AvailableDescriptors()
+    for key in _instance.__dict__.keys():
+        print(f"{key}")
+
+
+@requires(lds, "libdescriptor is needed for Descriptors")
+class Descriptor(ConfigurationTransform):
+    """
+    Descriptor class provides interface with the libdescriptor library. It provides a
+    unified interface to all the descriptors in libdescriptor, however all descriptors need
+    to have a corresponding initializer routine, to deal with hyperparameters. The descriptor
+    is initialized with a cutoff radius, a list of species,a descriptor type and an ordered
+    list of hyperparameters. The descriptor type is a string, which can be obtained by
+    `show_available_descriptors()` function. Some sane default values for the hyperparameters
+    are provided in ~:kliff.transforms.configuration_transforms.default_hyperparameters
+    module. This class methods to compute derivatives of the descriptor with respect to
+    atomic positions. The functions generating descriptors and their derivatives are
+    implemented as `forward()` and `backward()`, respectively, to match the PyTorch nomenclature.
+
+    """
+
+    def __init__(
+        self,
+        cutoff: float,
+        species: List[str],
+        descriptor: str,
+        hyperparameters: Union[Dict, str],
+        cutoff_function: str = "cos",
+        nl_ctx: NeighborList = None,
+        copy_to_config: bool = False,
+    ):
+        """
+        Args:
+            cutoff (float): Cutoff radius.
+            species (list): List of strings, each string is a species (atomic symbols).
+            descriptor (str): String of descriptor type, can be obtained by `show_available_descriptors()`.
+            hyperparameters (Dict): Ordered dictionary of hyperparameters.
+            cutoff_function (str): Cut-off function, currently only "cos" is supported.
+            nl_ctx (NeighborList): function to compute neighbor list, if not provided, will be computed internally.
+            copy_to_config (bool): If True, the fingerprint will be copied to
+                the Configuration object's fingerprint attribute.
+        """
+        super().__init__(copy_to_config)
+        self.cutoff = cutoff
+        self.species = species
+        _available_descriptors = AvailableDescriptors()
+        self.descriptor_name = descriptor
+        self.descriptor_kind = getattr(_available_descriptors, descriptor)
+        self.width = -1
+        self.hyperparameters = self.get_default_hyperparams(hyperparameters)
+        self.cutoff_function = cutoff_function
+        self._cdesc, self.width = self._init_descriptor_from_kind()
+        if nl_ctx:
```

Comment:
`Descriptor` is a subclass of `ConfigurationTransform`, so I expect `forward` and `backward` can be called multiple times, e.g. each time for a different configuration. 

However, `NeighborList` is bound to configuration, i.e. a `NeighborList` is needed for each config. 

So, I don't see how a single `nl_ctx=NeighborList` passed in __init__ can work for different configurations in `forward` and `backward`. It might be possible if one provides something else other than `NeighborList` to `nl_ctx`. Do we have that already? If not, I'd suggest removing `nl_ctx` to simplify things. If we want them, we can add back later.

Reply:
Ok. I think I know from where this bug crept up. It was orginally written as an torch.Autograd function thus the forward and backward were static function with independent `ctx` context being passed around. I modified it incorrectly to its current form. 

Edited it now to not have neighbor context.